packages/compass-import-export/src/import/analyze-csv-fields.ts

```diff
@@ -0,0 +1,240 @@
+  firstRowIndex: number;
+  firstColumnIndex: number;
+  firstValue: string;
+};
+
+// the subset of bson types that we can detect
+type CSVFieldType =
+  | 'int'
+  | 'long'
+  | 'double'
+  | 'boolean'
+  | 'date'
+  | 'string'
+  | 'null'
+  | 'mixed';
+
+/*
+For each field we need the detected types and the column positions. This helps
+with accounting for all column indexes, but also the fact that we'd have higher
+counts than the number of rows. ie. foo[0],foo[1] means twice as many fields as
+rows once it becomes field foo and so does foo[0].bar,foo[1].bar once it becomes
+foo.bar.
+*/
+type CSVField = {
+  types: Record<CSVFieldType, CSVFieldTypeInfo>;
+  columnIndexes: number[];
+  detected: CSVFieldType;
+};
+
+type AnalyzeCSVFieldsResult = {
+  totalRows: number;
+  fields: Record<string, CSVField>;
+};
+
+function detectFieldType(value: PapaValue): CSVFieldType {
+  // papaparse handles some types for us, so get those out of the way first
+  if (value === null) {
+    return 'null';
+  }
+  if (value instanceof Date) {
+    return 'date';
+  }
+  const jsType = typeof value;
+  if (jsType === 'boolean') {
+    return 'boolean';
+  }
+
+  // papaparse will make anything > -2^^53 and < 2^^53 a number (via parseFloat)
+  // and leave any number outside of that range a string.
+  if (jsType === 'number') {
+    // first separate floating point numbers from integers
+    if (value.toString().includes('.')) {
+      return 'double';
+    }
+    // then separate ints from longs
+    if (value >= MIN_INT && value <= MAX_INT) {
+      return 'int';
+    }
+    return 'long';
+  }
+
+  // now deal with remaining large ints
+  try {
+    const number = BigInt(value);
+    if (number >= MIN_LONG && number <= MAX_LONG) {
+      return 'long';
+    }
+    return 'string';
+  } catch (err) {
+    // If it doesn't parse as a BigInt, then just fall back to string.
+    // We might still want to check the string for certain patterns, though.
+    return 'string';
+  }
+}
+
+function initResultFields(
+  result: AnalyzeCSVFieldsResult,
+  headerFields: string[]
+) {
+  for (const [columnIndex, name] of headerFields.entries()) {
+    const fieldName = csvHeaderNameToFieldName(name);
+
+    if (!result.fields[fieldName]) {
+      result.fields[fieldName] = {
+        types: {} as Record<string, CSVFieldTypeInfo>,
+        columnIndexes: [],
+        detected: 'mixed', // we'll fill this at the end
+      };
+    }
+
+    // For fields inside arrays different CSV header fields can map to
+    // the same fieldName, so there will be more than one entry in
+    // columnIndex.
+    result.fields[fieldName].columnIndexes.push(columnIndex);
+  }
+}
+
+function addRowToResult(
+  result: AnalyzeCSVFieldsResult,
+  headerFields: string[],
+  data: PapaRowData,
+  currentRow: Record<string, string>
+) {
+  for (const field of Object.values(result.fields)) {
+    for (const columnIndex of field.columnIndexes) {
+      const fieldName = headerFields[columnIndex];
```

Comment:
ugh I just realised that I used the variable fieldName for two different things in two places which could be confusing so I'll rename them. in this one it is the field as it is in the csv header and in the other it is the field once it has been "normalised" ie. all array indexes removed.